src/Symfony/Component/Messenger/Middleware/HandleMessageMiddleware.php

```diff
@@ -41,12 +42,48 @@ public function handle(Envelope $envelope, StackInterface $stack): Envelope
         $handler = null;
         $message = $envelope->getMessage();
         foreach ($this->handlersLocator->getHandlers($envelope) as $handler) {
-            $handler($message);
+            $envelope = $envelope->with(new HandledStamp($handler($message), $this->formatCallable($handler)));
         }
         if (null === $handler && !$this->allowNoHandlers) {
             throw new NoHandlerForMessageException(sprintf('No handler for message "%s".', \get_class($envelope->getMessage())));
         }
 
         return $stack->next()->handle($envelope, $stack);
     }
+
+    private function formatCallable(callable $callable): ?string
```

Comment:
should be moved to the HandledStamp class to me.

Reply:
In practice, should not be an issue. At this point the envelope was already sent if async and would need to be serialized anymore.
Anyway the static factory would be fine, but what about the suggestion in description? 
I.e: the name would be yield as key from locators. This being the fallback if no name was provided. Would still live in the stamp to you?